File: 2020/day15.cpp
#include <iostream>
#include <vector>

using namespace std;

// run with:
// g++ -O3 day15.cpp; time ./a.out

int main() {
  int nums[] = {0,14,1,3,7,9};
  // int nums[] = {0,3,6};
  int size = sizeof(nums)/sizeof(nums[0]);
  vector<int> turns = vector<int>(100, -1);

  int last, num = 0;

  for (int i = 0; i < size; i++) {
    int n = nums[i];
    turns[n*2] = i;
    turns[n*2+1] = 0;
    last = n;
  }

  for (int i = size; i < 30000000; i++) {
    if (turns.size() < last*2+1) {
      turns.resize(last*2+1, -1);
    }
    num = turns[last*2+1];
    if (turns.size() < num*2+1) {
      turns.resize(num*2+1, -1);
    }
    if (num == 0 || turns[num*2] > 0) {
      turns[num*2+1] = i - turns[num*2];
      turns[num*2] = i;
    } else {
      turns[num*2] = i;
      turns[num*2+1] = 0;
    }
    last = num;
  }

  cout << num << endl;
}

File: 2020/day15.py
def run(filename):
  with open(filename, 'r') as f:
    nums = [int(n) for n in f.readline().split(',')]
    turns = {}
    for i in range(len(nums)):
      n = nums[i]
      turns[n] = [i, 0]

    last = nums[-1]
    num = 0

    for i in range(len(nums), 30000000):
      num = turns[last][1]

      if num in turns:
        turns[num][1] = i - turns[num][0]
        turns[num][0] = i
      else:
        turns[num] = [i, 0]

      last = num
      if i % 1000000 == 0:
        print(i)
    print(num)


# run('day15_ex.txt')
run('day15.txt') # 1876406